File: api/notion.py
import os
from notion_client import Client

notion = Client(auth=os.environ["NOTION_TOKEN"])


def get_database():
    output = {}

    # Query database
    results = notion.databases.query(
        **{
            "database_id": os.environ["NOTION_DATABASE_ID"],
            "filter": {
                "property": "Owned",
                "checkbox": {
                    "equals": False
                }
            }
        }
    )['results']

    # Iterate through rows
    for row in results:
        props = row['properties']
        row_output = {}

        # Extract the title for the key
        title_property = next(
            (value for key, value in props.items() if value['type'] == 'title'), None
        )
        if not title_property:
            continue  # Skip rows without a title property

        # Extract the title's text content
        page_title = ''.join(
            [text['text']['content'] for text in title_property['title']]
        )

        # Process each property dynamically
        for key, value in props.items():
            _type = value['type']  # Get the property type
            prop_value = value[_type]

            # Extract values based on property type
            if _type == "checkbox":
                row_output[key] = prop_value
            elif _type == "number":
                row_output[key] = prop_value
            elif _type == "select":
                row_output[key] = prop_value['name'] if prop_value else None
            elif _type == "rich_text":
                row_output[key] = ''.join([text['text']['content'] for text in prop_value])
            elif _type == "url":
                row_output[key] = prop_value
            elif _type == "title":
                row_output[key] = ''.join([text['text']['content'] for text in prop_value])
            else:
                row_output[key] = None  # Handle unsupported types

        # Add row to output with page title as the key
        output[page_title] = row_output

    print(output)

    return output

File: lib/notion.ts
import { Client } from "@notionhq/client";
import { DatabaseResponse, NotionProperty } from "./types";
import { PageObjectResponse, PartialPageObjectResponse, DatabaseObjectResponse } from "@notionhq/client/build/src/api-endpoints";

const notion = new Client({
  auth: process.env.NOTION_TOKEN,
});

export async function getDatabase(): Promise<DatabaseResponse> {
  const output: DatabaseResponse = {};
  const results = await notion.databases.query({
    database_id: process.env.NOTION_DATABASE_ID!,
    filter: {
      property: "Owned",
      checkbox: {
        equals: false,
      },
    },
  });

  for (const row of results.results) {
    // Type guard to check if this is a full page object
    if (!('properties' in row)) continue;

    const props = (row as PageObjectResponse).properties as Record<string, NotionProperty>;
    const rowOutput = {
      Name: "",
      Brand: "",
      Category: "",
      Price: 0,
      "Purchase Link": "",
      PageBanner: null as string | null,
    };

    // Find title property
    const titleProperty = Object.values(props).find(
      (value) => value.type === "title"
    );
    if (!titleProperty) continue;

    // Extract title
    const pageTitle = titleProperty.title
      .map((text: any) => text.text.content)
      .join("");

    // Get page cover/banner if exists
    const pageCover = (row as PageObjectResponse).cover;
    rowOutput.PageBanner = pageCover?.type === "external"
      ? pageCover.external.url
      : pageCover?.type === "file"
        ? pageCover.file.url
        : null;

    // Process properties
    for (const [key, value] of Object.entries(props)) {
      const type = value.type;
      const propValue = value[type];

      switch (type) {
        case "checkbox":
          rowOutput[key] = propValue as boolean;
          break;
        case "number":
          rowOutput[key] = propValue as number;
          break;
        case "select":
          rowOutput[key] = (propValue as { name: string } | null)?.name ?? null;
          break;
        case "rich_text":
          rowOutput[key] = (propValue as Array<{ text: { content: string } }>)
            .map((text) => text.text.content)
            .join("");
          break;
        case "url":
          rowOutput[key] = propValue as string;
          break;
        case "title":
          rowOutput[key] = (propValue as Array<{ text: { content: string } }>)
            .map((text) => text.text.content)
            .join("");
          break;
        default:
          rowOutput[key] = null;
      }
    }

    output[pageTitle] = rowOutput;
  }

  return output;
}